lib/mod_movements/virtualWorld.js

```diff
@@ -0,0 +1,76 @@
+/**
+ * Checks to make sure two positions are equal.
+ *
+ * @param {Vec3} a - The first position
+ * @param {Vec3} b - The second position
+ */
+function samePos (a, b) {
+  return a.x === b.x && a.y === b.y && a.z === b.z
+}
+
+/**
+ * A wrapper for the block world which allows for checking what the world state would
+ * be along a theoretical path.
+ */
+class VirtualWorld {
+  constructor (bot, world) {
+    this.bot = bot
+    this.world = world
+  }
+
+  /**
+   * Gets what block would be at this location at this node along the path.
+   *
+   * @param {Movement} node - The current movement node to check.
+   * @param {*} position - The position within the world.
+   */
+  virtualBlockAt (node, position) {
```

Comment:
I get what you want to do here, but I don't think it is worth it. The current algorithm already assume that the position the block is standing in is valid and most of the time it will not revisit previous nodes. In addition, paths are recomputed each time the environment changes (including when the bot modifies it). The block access needs to be kept fast, so this is not something you want to change unless there is a very good reason for it.

Can you provide an example where taking past actions into account gives a better quality path than what is currently implemented ?

Reply:
Yes, a common example of this, including addressing the configurable hashing implementation, would be for world-editing based tasks. Namely, where the goal of the task is not to get to a certain position but to edit a certain number of blocks. Storing what blocks where modified and how is essential.

Another example is used when handling combined goals where two objectives must be met. This may include breaking a block to access one of the waypoints and replacing it to bridge across to another.

One thing that concerns me is the bot recalculating the path when modifying it's own blocks. This leads to the bot "forgetting" information such as waypoints like mentioned above.

At current, the bot is only capable of moving from one destination to another with little room for configuring extra actions to be preformed along the way.

Reply:
For both cases, it is more efficient to address the issue at a higher level. When I was asking for an example, I meant a real (implemented) example on which you run both versions and compare the output.

In other words, the waypoint and building use cases are both very interesting, but they are out of scope of the modular movement system (and there is more than one way to implement them). IMO there is no need to complicate and slow things now for features we don't need yet.